### PR TITLE
NDEV-1723: Simplify plugin

### PR DESCRIPTION
**Pull request by andreisilviudragnea**
- Simplified Kafka threading model by using only one Kafka `ThreadedProducer` instead of 4 `FutureProducer`s
- Used type state pattern for managing plugin states
- Restructured plugin configuration JSON to a nested structure
- Redefined most plugin Prometheus metrics around the Kafka producer `Statistics`
- Fixed plugin graceful shutdown
- Added unit test for loading and unloading the plugin

I added a more detailed explanation for each bullet here: https://www.notion.so/neonlabs/Geyser-Neon-plugin-updates-2d1d0477e0344e689b168048f9270bb7

### CODE REVIEW

File: kafka_common/src/hash.rs [hunk 1]
```diff
@@ -16,27 +22,21 @@ impl UpdateAccount {
             }
         }
 
-        hasher.update(&self.slot.to_le_bytes());
+        hasher.update(&self.slot.to_le_bytes()); // TODO Why is this included two times?
```
Comment (Deniskore) [inline kafka_common/src/hash.rs:25]: It's my typo. We should remove it.

File: kafka_common/src/kafka_impl.rs [hunk 2]
```diff
@@ -127,7 +127,7 @@ impl From<&ReplicaTransactionInfoV2<'_>> for KafkaReplicaTransactionInfoV2 {
 impl UpdateAccount {
     pub fn set_write_version(&mut self, transaction_index: i64) {
         let mut account = match &mut self.account {
-            KafkaReplicaAccountInfoVersions::V0_0_1(_) => unreachable!(),
+            KafkaReplicaAccountInfoVersions::V0_0_1(_) => unreachable!(), // TODO: Why only for V2? V1 also has write_version
```
Comment (Deniskore) [inline kafka_common/src/kafka_impl.rs:130]: V1 only for validator version <= 1.13.x
For 1.14 >= the version is V0_0_2
We use 1.14 (V0_0_2)

**Review by Deniskore — APPROVED** [review]
LGTM.

File: kafka_common/src/hash.rs (hunk 1 above)
Comment (dk-neon) [reply kafka_common/src/hash.rs]: Why don't remove now?

**Review by dk-neon — CHANGES_REQUESTED** [review]

File: kafka_common/src/hash.rs (hunk 1 above)
Comment (andreisilviudragnea) [reply kafka_common/src/hash.rs]: I removed it

**Review by dk-neon — APPROVED** [review]
LGTM

**Pull request merged**